coretrace/simulation_data/cst_templates/heliostat.cpp

```diff
@@ -0,0 +1,437 @@
+    panel_len_x /= this->num_panels_x;
+    double panel_len_y = this->aperture_size_y -
+                         this->gap_y * (this->num_panels_y - 1);
+    panel_len_y /= this->num_panels_y;
+
+    if (this->canting_method == OFF_AXIS)
+    {
+        // TODO: Do stuff here...
+    }
+
+    this->heliostat_area = 0.0;
+    double panel_x = -0.5 * (this->aperture_size_x - panel_len_x);
+    double panel_y = -0.5 * (this->aperture_size_y - panel_len_y);
+
+    single_element_ptr elem;
+    Vector3d origin;
+    Vector3d aim;
+    aperture_ptr ap;
+    surface_ptr surf;
+    element_id sts;
+    double c, z;
+
+    for (int j = 0; j < this->num_panels_y; ++j)
+    {
+        for (int i = 0; i < this->num_panels_x; ++i)
+        {
+            elem = make_element<SingleElement>();
+
+            if (this->canting_method == NONE)
+            {
+                origin.set_values(panel_x, panel_y, 0.0);
+                aim.set_values(panel_x, panel_y, 1000.0);
+            }
+            else if (this->canting_method == ON_AXIS)
+            {
+                c = 0.5 / (this->onaxis_canting_distance);
+                z = 0.5 * c * (panel_x * panel_x + panel_y * panel_y);
+                origin.set_values(panel_x, panel_y, z);
+                aim.set_values(0.0, 0.0, 2.0 * this->onaxis_canting_distance);
+            }
+            else if (this->canting_method == OFF_AXIS)
+            {
+                origin.set_values(panel_x, panel_y, 0.0);
+                // TODO: Set aim vector values
+                aim.set_values(0.0, 0.0, 1.0);
+                throw std::runtime_error("OFF_AXIS is not yet implemented");
+            }
+            else if (this->canting_method == UNSET)
+            {
+                throw std::runtime_error("Canting method unset");
+            }
+            else
+            {
+                throw std::runtime_error("Unknown canting method");
+            }
+
+            ap = make_aperture<Rectangle>(panel_len_x, panel_len_y);
+            elem->set_aperture(ap);
+
+            // // TODO: Is this the correct thing to do here?
+            // vector_add(1.0, this->focal_point, -1.0, origin, aim);
+
+            if (this->focal_length_x <= 0.0 && this->focal_length_y <= 0.0)
+            {
+                surf = make_surface<Flat>();
+                // Vector3d khat(0.0, 0.0, 1.0);
+                // vector_add(0.5, khat, 0.5, aim);
+            }
+            else
+            {
+                surf = make_surface<Parabola>(this->focal_length_x,
+                                              this->focal_length_y);
+            }
+            elem->set_surface(surf);
+
+            elem->set_reference_frame_geometry(origin,
+                                               aim,
+                                               0.0);
+
+            elem->set_front_optical_properties(this->optics_mirror);
+            elem->set_back_optical_properties(this->optics_mirror);
```

Comment:
We should probably make the back surface properties different. Either fixed to some default "absorber" or provide access to customize.

Reply:
I will add a TODO about this.

It was while doing these templates that I realized we should rework the optical properties to have be a single object with a front and back so we can do error checking make it a bit easier to create elements (less verbose).